manim/utils/tex.py

```diff
@@ -1,298 +1,189 @@
-            "documentclass",
-            required_params=[self.documentclass[0]],
-            optional_params=self.documentclass[1],
+        self.output_format = (
+            output_format
+            if output_format is not None
+            else TexTemplate.default_output_format
+        )
+        self.documentclass = (
+            documentclass
+            if documentclass is not None
+            else TexTemplate.default_documentclass
+        )
+        self.preamble = (
+            preamble if preamble is not None else TexTemplate.default_preamble
+        )
+        self.placeholder_text = (
+            placeholder_text
+            if placeholder_text is not None
+            else TexTemplate.default_placeholder_text
+        )
+        self.post_doc_commands = (
+            post_doc_commands
+            if post_doc_commands is not None
+            else TexTemplate.default_post_doc_commands
+        )
+        self.rebuild()
+
+    def rebuild(self):
+        self.body = (
+            self.documentclass
+            + "\n"
+            + self.preamble
+            + "\n"
+            + r"\begin{document}"
+            + "\n"
+            + self.post_doc_commands
+            + "\n"
+            + self.placeholder_text
+            + "\n"
+            + r"\end{document}"
+            + "\n"
         )
-        for pkg in self.common_packages:
-            tpl += self.generate_usepackage(pkg)
-
-        if self.use_ctex:
-            for pkg in self.ctex_packages:
-                tpl += self.generate_usepackage(pkg)
-        else:
-            for pkg in self.tex_packages:
-                tpl += self.generate_usepackage(pkg)
-
-        tpl += self.common_preamble_text
-        if self.use_ctex:
-            tpl += self.ctex_preamble_text
-        else:
-            tpl += self.tex_preamble_text
-
-        tpl += "\n" r"\begin{document}" "\n"
-        tpl += f"\n{self.text_to_replace}\n"
-        tpl += "\n" r"\end{document}"
-
-        self.body = tpl
 
-    def prepend_package(self, pkg):
-        """Adds a new package (or several new packages)
-        before all other packages. Sometimes, the order of
-        the ``\\usepackage`` directives is relevant.
+    def prepend_to_preamble(self, txt):
+        """Adds txt to the TeX template preamble just after the \documentclass
 
         Parameters
         ----------
-        pkg : :class:`str`
-            The package name, e.g. ``"siunitx"``
+        txt : :class:`str`
+            String containing the text to be added, e.g. ``\\usepackage{siunitx}``
         """
-        self.common_packages.insert(0, pkg)
-        self.rebuild_cache()
+        self.preamble = txt + "\n" + self.preamble
+        self.rebuild()
 
-    def append_package(self, pkg):
-        """Adds a new package (or several new packages)
-        after all other packages. Sometimes, the order of
-        the ``\\usepackage`` directives is relevant.
+    def add_to_preamble(self, txt):
```

Comment:
```suggestion
    def add_to_preamble(self, txt, prepend=False):
        """Adds stuff to the TeX template's preamble (e.g. definitions, packages). Text can be inserted at the beginning or at the end of the preamble.

        Parameters
        ----------
        txt : :class:`str`
            String containing the text to be added, e.g. ``\\usepackage{hyperref}``
        prepend : Optional[:class:`bool`], optional
            Whether the text should be added at the beginning of the preample, i.e. right after ``\\documentclass``. Default is to add it at the end of the preample, i.e. right before ``\\begin{document}``
        """
        if prepend:
            self.preamble = txt + "\n" + self.preamble
        else:
            self.preamble += "\n" + txt
        self.rebuild()
```

Liked your suggestion of having just one method for both cases. Sorry, review comments are a bit messy, but it seems impossible to make a suggestion when there are deleted lines in between.